fix: add netns@podns.service dependency to setup-nat-for-imds

gsid100 opened the pull request:
The `setup-nat-for-imds` service was failing to start because it attempted to access the `podns` network namespace before it was created.

This change explicitly adds a dependency on `netns@podns.service` to ensure the namespace is available before `setup-nat-for-imds` runs.

Fixes: #2385

gsid100 commented:
/hold

vvoronko commented:
/lgtm

ldoktor commented on src/cloud-api-adaptor/podvm/files/etc/systemd/system/setup-nat-for-imds.service:
```diff
@@ -1,8 +1,8 @@
-After=agent-protocol-forwarder.service
-Wants=agent-protocol-forwarder.service
+After=agent-protocol-forwarder.service netns@podns.service
+Wants=agent-protocol-forwarder.service netns@podns.service
```
This makes sense to me, but is this sufficient? Shouldn't we add a check that the netns is already up and prepared to be configured? (I tried it locally and even with the After it's failing, but with extra `sleep 10` it's passing).

I tried adding `ip netns exec "$NETNS" ping -c1 -W1 127.0.0.1 &>/dev/null` but that did not help, not really a networking guy so do you have any ideas?

gsid100 replied:
Hey, this is correct, the podns wasn't the missing piece for the setup-nat service.
It's still unclear to me what was missing in the network configuration, but during the initial attempt to start the setup-nat service the environment isn't ready yet. I added a restart on failure policy to the service file, things are working well now.

gsid100 commented:
My reply to @ldoktor:
the podns wasn't the missing piece for the setup-nat service.
It's still unclear to me what was missing in the network configuration, but during the initial attempt to start the setup-nat service the environment isn't ready yet. I added a restart on failure policy to the service file, things are working well now.

I updated this PR, would appreciate a review

Gilad

mkulke reviewed (commented):
this probably works, but do we know why we have to use restarts and delays?

vvoronko commented:
/lgtm

gsid100 commented:
> this probably works, but do we know why we have to use restarts and delays?

I haven't pinpointed the exact dependency yet, but it seems that not all networking commands are ready for use immediately when the service starts up

ldoktor reviewed (changes requested):
From my testing it doesn't look like the script is prepared to be re-executed, which might (and does) results in even worse issues:

```
May 21 09:19:11 fedora systemd[1]: Starting setup-nat-for-imds.service - Setup NAT for IMDS...
May 21 09:19:11 fedora systemd-networkd[460]: veth1: Link UP
May 21 09:19:11 fedora setup-nat-for-imds.sh[1063]: net.ipv4.ip_forward = 1
May 21 09:19:11 fedora setup-nat-for-imds.sh[1067]: net.ipv4.conf.veth1.proxy_arp = 1
May 21 09:19:11 fedora setup-nat-for-imds.sh[1068]: net.ipv4.neigh.veth1.proxy_delay = 0
May 21 09:19:11 fedora systemd-networkd[460]: veth1: Gained carrier
May 21 09:19:11 fedora systemd[1]: setup-nat-for-imds.service: Deactivated successfully.
May 21 09:19:11 fedora systemd[1]: Finished setup-nat-for-imds.service - Setup NAT for IMDS.
May 21 09:19:11 fedora systemd[1]: Startup finished in 1.095s (kernel) + 1.716s (initrd) + 17.486s (userspace) = 20.298s.
May 21 09:19:13 fedora systemd-networkd[460]: veth1: Gained IPv6LL
May 21 09:19:37 fedora systemd[1]: systemd-hostnamed.service: Deactivated successfully.
May 21 09:20:23 fedora systemd[1]: Starting netns@podns.service...
May 21 09:20:23 fedora ip[1103]: Cannot create namespace file "/var/run/netns/podns": File exists
May 21 09:20:23 fedora systemd[1]: netns@podns.service: Control process exited, code=exited, status=1/FAILURE
May 21 09:20:23 fedora systemd[1]: netns@podns.service: Failed with result 'exit-code'.
May 21 09:20:23 fedora systemd[1]: Failed to start netns@podns.service.
```

In general I'm not against the re-try, makes sense to me, but either we need to modify the script to allow re-execution or we just have to detect the system is ready to be configured prior running our setup.

mkulke commented:
I would also prefer to understand the issue before we solve it with delays and retries. Maybe there is a condition that we can wait for, or we have to start it after the kata-agent unit has launched

ldoktor commented:
How about something like this (only checked it in a running pod, need to build an image to see if it really  behaves correctly)

```diff
diff --git a/src/cloud-api-adaptor/podvm/files/usr/local/bin/setup-nat-for-imds.sh b/src/cloud-api-adaptor/podvm/files/usr/local/bin/setup-nat-for-imds.sh
index 34b0f6d3..25383f26 100755
--- a/src/cloud-api-adaptor/podvm/files/usr/local/bin/setup-nat-for-imds.sh
+++ b/src/cloud-api-adaptor/podvm/files/usr/local/bin/setup-nat-for-imds.sh
@@ -1,43 +1,72 @@
 #!/bin/bash
 # This script sets up NAT for IMDS on Azure/AWS
-# This is required for IMDS to work on Azure/AWS
-# This script is executed as a oneshot systemd service
-# during first boot
+# It can be safely re-executed to complete or fix a partial setup.
 
 set -euo pipefail
 
 IMDS_IP="169.254.169.254"
 DUMMY_IP="169.254.99.99"
+VETH_HOST="veth1"
+VETH_NS="veth2"
+NETNS="podns"
 
 # trap errors
 trap 'echo "Error: $0:$LINENO stopped"; exit 1' ERR INT
 
-# Function to setup veth pair
+# Function to check if a network device exists
+function link_exists() {
+  ip link show "$1" &>/dev/null
+}
+
+# Function to setup veth pair and routing/NAT
 function setup_proxy_arp() {
   local pod_ip
-  pod_ip=$(ip netns exec podns ip route get "$IMDS_IP" | awk '{ for(i=1; i<=NF; i++) { if($i == "src") { print $(i+1); break; } } }')
+  pod_ip=$(ip netns exec "$NETNS" ip route get "$IMDS_IP" | awk '{ for(i=1; i<=NF; i++) { if($i == "src") { print $(i+1); break; } } }')
+
+  # Create veth pair only if it doesn't exist
+  if ! link_exists "$VETH_HOST"; then
+    ip link add "$VETH_NS" type veth peer name "$VETH_HOST"
+  fi
+
+  # Assign IP to veth1 if not already assigned
+  if ! ip address show dev "$VETH_HOST" | grep -q "$DUMMY_IP"; then
+    ip address add "$DUMMY_IP/32" dev "$VETH_HOST"
+  fi
 -  ip link add veth2 type veth peer name veth1
-  # Proxy arp does not get enabled when no IP address is assigned
-  ip address add "$DUMMY_IP/32" dev veth1
-  ip link set up dev veth1
+  ip link set up dev "$VETH_HOST"
 
-  sysctl -w net.ipv4.ip_forward=1
-  sysctl -w net.ipv4.conf.veth1.proxy_arp=1
-  sysctl -w net.ipv4.neigh.veth1.proxy_delay=0
+  sysctl -w net.ipv4.ip_forward=1 >/dev/null
+  sysctl -w net.ipv4.conf."$VETH_HOST".proxy_arp=1 >/dev/null
+  sysctl -w net.ipv4.neigh."$VETH_HOST".proxy_delay=0 >/dev/null
 
-  ip link set veth2 netns podns
-  ip netns exec podns ip link set up dev veth2
-  ip netns exec podns ip route add "$IMDS_IP/32" dev veth2
+  # Move veth2 into the pod namespace if not already
+  if ! ip netns exec "$NETNS" ip link show "$VETH_NS" &>/dev/null; then
+    ip link set "$VETH_NS" netns "$NETNS"
+  fi
 
-  ip route add "$pod_ip/32" dev veth1
+  ip netns exec "$NETNS" ip link set up dev "$VETH_NS"
 
+  # Add route to IMDS via veth2 if not present
+  if ! ip netns exec "$NETNS" ip route show | grep -q "$IMDS_IP"; then
+    ip netns exec "$NETNS" ip route add "$IMDS_IP/32" dev "$VETH_NS"
+  fi
+
+  # Route pod IP via veth1
+  if ! ip route get "$pod_ip" | grep -q "dev $VETH_HOST"; then
+    ip route add "$pod_ip/32" dev "$VETH_HOST"
+  fi
+
+  # Set static ARP entry
   local hwaddr
-  hwaddr=$(ip netns exec podns ip -br link show veth2 | awk 'NR==1 { print $3 }')
-  ip neigh replace "$pod_ip" dev veth1 lladdr "$hwaddr"
+  hwaddr=$(ip netns exec "$NETNS" ip -br link show "$VETH_NS" | awk 'NR==1 { print $3 }')
+  ip neigh replace "$pod_ip" dev "$VETH_HOST" lladdr "$hwaddr"
 
-  iptables -t nat -A POSTROUTING -s "$pod_ip/32" -d "$IMDS_IP/32" -j MASQUERADE
+  # Add iptables NAT rule only if not present
+  if ! iptables -t nat -C POSTROUTING -s "$pod_ip/32" -d "$IMDS_IP/32" -j MASQUERADE 2>/dev/null; then
+    iptables -t nat -A POSTROUTING -s "$pod_ip/32" -d "$IMDS_IP/32" -j MASQUERADE
+  fi
 }

# Execute function
setup_proxy_arp
```

ldoktor commented:
> I would also prefer to understand the issue before we solve it with delays and retries. Maybe there is a condition that we can wait for, or we have to start it after the kata-agent unit has launched

Well we could additionally use `ip netns list` to see if the ns is created, but not sure it's sufficient as from my attempts created != usable.

mkulke commented:
sure, we can make this script idempotent and perform systemd retries, but this is not exactly pretty. e.g. the boot logs will look messy. isn't there a systemd condition that we can wait for? this looks to me like it would be required very late in the boot process, after kata-agent.

 If there really is no such condition, can we put the wait loop inside the script?

ldoktor commented:
That's something I played with before the retries, but the wait for ns wasn't sufficient. I think it'd be nice to make it idempotent and put the ns wait at the beginning (for up to given time and then fail or proceed). That way we get nicer logs most of the time and if retry happens it should be safe.

ldoktor commented:
Btw I functionally tested the idempotent version I shared earlier and it seems to be working well.

ldoktor commented:
https://github.com/confidential-containers/cloud-api-adaptor/pull/2450 was merged instead
